### PR TITLE
Extend fpfss access token into extensions

### PR DESCRIPTION
**Pull request by dot-mike**
This PR introduces the `getAccessToken()` method in a new flashpoint.fpfss namespace, allowing extensions to get access token for the logged in user with FPFSS.

The way it works is that the frontend will still handle the calls to FPFSS and the backend acts like a proxy between the extension and the frontend ~~using a new state `pendingFpfssActions` to keep track of calls between frontend and backend.~~

So even if the user is logged out and the extensions call `getAccessToken()` the main render process will trigger a login to FPFSS by opening the web browser.

**NOTE** there is no warning here, the web browser will open immediately asking for user permission

EDIT 1:
Consent mechanism implemented as discussed. Only the first time the extensions request access to FPFSS the user will be prompted to allow access. The access can be revoked in the config page.

~~I think we really should split this `performFpfssAction` into two methods, one for internal application stuff and another one for extensions. By allowing for extensions access `performFpfssAction` through `getAccessToken` without any form of checks we are essentially opening up for credential stealing of the bearer token.~~

~~For this reasion I don't want _any_ extensions by default to be allowed to your access token without consent the first time the extensions tries to call this method. Thoughts on this?~~

### CODE REVIEW

**Comment by colin969** [comment]
Looks good for the most part.

I agree on the security considerations. I'd pass the extension ID along with the request to the frontend, then compare against a list of approved extensions kept in localstorage. It can prompt with a dialog. Much less open to malicious intent.

You don't need a separate return action over the websocket, it's smart enough that `register` can accept async functions and `request` can be awaited so you can use them like any other async function. This is a bit confusing, because there isn't an actual event for you to respond to (normally you'd grab the client from the event), but you can still get the connected client from `state.socketServer.lastClient`. The backend's SocketServer can pass the client as the first argument of a `send` or `request` call. Just remember `lastClient` can be null if a renderer isn't connected, in which case you just have to back off.

This is a bit on me, since the perform function is using callbacks to begin with. It should be moved to a proper async / await function which returns the user rather than being given a callback, then have the other uses of it modified to gracefully handle it.

**Comment by colin969** [comment]
Moved the base to release branch, has conflicts which Github consider too big to show, but it's probably just the diff generator getting upset about the placement of your function registering, shouldn't be anything major.

**Comment by dot-mike** [comment]
Thanks for your knowledge and vaulable feedback! I just tested the changes without having a separate return over the websocket and I can confirm it works! That is just awesome 👍

**Comment by dot-mike** [comment]
I've updated so the user has to approve FPFSS usage for extensions. State machine in diagram.

```mermaid
stateDiagram-v2
    [*] --> ButtonClicked
    ButtonClicked: User clicks action button
    ButtonClicked --> Approved: fpfssConsent = true
    Approved --> PerformAction: Action proceeds as normal
    PerformAction --> [*]: Action completed successfully
    ButtonClicked --> AwaitingConsent: fpfssConsent = false
    AwaitingConsent --> Denied: User rejects consent
    Denied: Save to local storage as denied
    Denied --> Config: User must approve consent in configuration
    AwaitingConsent --> Approved: User approves consent in launcher
```

File: src/renderer/app.tsx [hunk 1]
```diff
@@ -672,8 +675,47 @@ export class App extends React.Component<AppProps> {
...
+          this.performFpfssAction(async (user) => {
+            resolve(user);
+          });
+        } else if (previousConsent === false) {
+          reject(new Error('Extension is not allowed to access FPFSS token'));
+        } else {
+          remote.dialog.showMessageBox({
+            type: 'question',
+            title: 'FPFSS Extension Access',
+            message: `Extension with ID ${extId} is requesting access to your FPFSS token. Allow?`,
```
Comment (dot-mike) [inline src/renderer/app.tsx:692]: Need to use translation lookup here.

File: src/renderer/app.tsx [hunk 2]
```diff
@@ -672,8 +675,47 @@ export class App extends React.Component<AppProps> {
...
+            cancelId: 1,
+          })
+          .then(({ response }) => {
+            if (response === 0) {
+              this.performFpfssAction(async (user) => {
+                if (user) {
+                  saveFpfssConsentExt(extId, true);
+                  resolve(user);
+                } else {
+                  saveFpfssConsentExt(extId, false);
```
Comment (dot-mike) [inline src/renderer/app.tsx:704]: We shouldn't save that the user didn't allow launcher to access FPFSS.
@colin969 agree?

Comment (colin969) [reply src/renderer/app.tsx]: Probably for the best only to save approvals. If an extension repeatedly asks for permission and the user won't allow it, then it should be up to the extension to behave and the user can remove them.

**Comment by dot-mike** [comment]
I believe this is ready for merge now!

**Pull request merged**